File: solutions/java/rail-fence-cipher/1/src/main/java/RailFenceCipher.java
class RailFenceCipher {
    private final int rows;

    RailFenceCipher(int rows) {
        this.rows = rows;
    }

    String getEncryptedData(String message) {
        // A fence with 1 or fewer rails doesn't change the message
        if (rows <= 1) {
            return message;
        }
        
        StringBuilder[] fences = new StringBuilder[rows];
        for (int i = 0; i < rows; i++) {
            fences[i] = new StringBuilder();
        }
        
        int currentRow = 0;
        int direction = 1; // 1 for moving down, -1 for moving up
        
        // Simulate the zig-zag pattern to distribute characters
        for (char c : message.toCharArray()) {
            fences[currentRow].append(c);
            
            // Change direction at the top and bottom rails
            if (currentRow == 0) {
                direction = 1;
            } else if (currentRow == rows - 1) {
                direction = -1;
            }
            currentRow += direction;
        }
        
        // Concatenate all rows to form the ciphertext
        StringBuilder result = new StringBuilder();
        for (StringBuilder fence : fences) {
            result.append(fence);
        }
        
        return result.toString();
    }

    String getDecryptedData(String message) {
        if (rows <= 1) {
            return message;
        }
        
        int n = message.length();
        int[] rowLengths = new int[rows];
        
        int currentRow = 0;
        int direction = 1;
        
        // Step 1: Figure out how many characters belong to each row
        for (int i = 0; i < n; i++) {
            rowLengths[currentRow]++;
            if (currentRow == 0) {
                direction = 1;
            } else if (currentRow == rows - 1) {
                direction = -1;
            }
            currentRow += direction;
        }
        
        // Step 2: Split the ciphertext into its respective rows
        String[] rowStrings = new String[rows];
        int index = 0;
        for (int i = 0; i < rows; i++) {
            rowStrings[i] = message.substring(index, index + rowLengths[i]);
            index += rowLengths[i];
        }
        
        // Step 3: Read the characters back in the original zig-zag order
        int[] rowIndices = new int[rows]; // Tracks which character to pull next from each row
        StringBuilder result = new StringBuilder();
        currentRow = 0;
        direction = 1;
        
        for (int i = 0; i < n; i++) {
            result.append(rowStrings[currentRow].charAt(rowIndices[currentRow]));
            rowIndices[currentRow]++;
            
            if (currentRow == 0) {
                direction = 1;
            } else if (currentRow == rows - 1) {
                direction = -1;
            }
            currentRow += direction;
        }
        
        return result.toString();
    }
}

File: solutions/javascript/rail-fence-cipher/1/rail-fence-cipher.js
export const encode = (text, numRails) => {
  if (numRails < 2) return text;
  
  const rails = Array.from({ length: numRails }, () => []);
  let currentRail = 0;
  let direction = 1;

  for (const char of text) {
    rails[currentRail].push(char);
    
    // Move to the next rail
    currentRail += direction;
    
    // Reverse direction if we hit the top or bottom rail
    if (currentRail === 0 || currentRail === numRails - 1) {
      direction *= -1;
    }
  }

  // Join all rails into a single ciphertext string
  return rails.map(rail => rail.join('')).join('');
};

export const decode = (text, numRails) => {
  if (numRails < 2) return text;
  
  // Step 1: Calculate the exact length of each rail
  const railLengths = new Array(numRails).fill(0);
  let currentRail = 0;
  let direction = 1;

  for (let i = 0; i < text.length; i++) {
    railLengths[currentRail]++;
    
    currentRail += direction;
    if (currentRail === 0 || currentRail === numRails - 1) {
      direction *= -1;
    }
  }

  // Step 2: Slice the ciphertext into the reconstructed rails
  const rails = [];
  let pointer = 0;
  for (let i = 0; i < numRails; i++) {
    // We split the slice into an array so we can use .shift() later
    rails.push(text.slice(pointer, pointer + railLengths[i]).split(''));
    pointer += railLengths[i];
  }

  // Step 3: Reconstruct the original text by reading out the zig-zag
  let decoded = '';
  currentRail = 0;
  direction = 1;

  for (let i = 0; i < text.length; i++) {
    decoded += rails[currentRail].shift();
    
    currentRail += direction;
    if (currentRail === 0 || currentRail === numRails - 1) {
      direction *= -1;
    }
  }

  return decoded;
};
